xosview.1
.TH "XOSVIEW" 1
.UC
.SH NAME
xosview \- X based system monitor
.SH SYNOPSIS
xosview [options]
.\"  ================  Macros  =============================
.\"  First, let's define some handy roff macros.
.\"  A macro begins with .de <xx> where one will invoke this macro with .xx
.\"  The macro definition ends with the .. line.
.\"  I don't know what macro abbreviations are free, so I just chose a few,
.\"  and haven't noticed a problem so far!  bgrayson

.\"  There are several paragraphs that are repeated in the resource section.
.\"  Rather than typing the whole stuff out each time, we define a few macros.

.\"  The .pp macro takes a single argument (net, disk, etc), and
.\" prints a paragraph description of a Priority resource.  Only the
.\" header (xosview*diskPriority: \fIpriority\fP) needs to be specified in
.\" addition to the .pp macro.
.\"  Usage:   .pp load
.de pp
xosview*\\$1Priority: \fIpriority\fP
.RS
This number (which must be an integer >= 1) sets the number of tenths of
a second that the \\$1 meter waits between updates.  A value of 1 has xosview
update the meter 10 times per second (the fastest).  A value of 600 would
cause xosview to update the meter once a minute.
.RE
..
.\"  The .dc macro is similar to the .pp macro, except that it is for
.\" the decay resource paragraphs.
.\"  Usage:  .dc net
.de dc
xosview*\\$1Decay: (True or False)
.RS
If True then the \\$1 meter will be split vertically in two.  The top
half will show the instantaneous state, while the bottom half will
display a decaying average of the state.
.RE
..
.\" The .dg macro is similar to the .dc macro, except that it is for
.\" the scrolling graph resource paragraphs.
.\" Usage:  .dg net
.de dg
xosview*\\$1Graph: (True or False)
.RS
If this is set to True then the \\$1 meter will be drawn as a horizontally
scrolling bargraph showing the state value verses time.
.RE
..
.\"  The .pm macro is for ``plus/minus'' -- for the
.\"  enable/disable command-line options.
.de pm
-\\$1
.RS
This option overrides the xosview*\\$1 resource.  It is
equivalent to setting xosview*\\$1 to "False".
.RE

+\\$1
.RS
This option overrides the xosview*\\$1 resource.  It is
equivalent to setting xosview*\\$1 to "True".
.RE
..
.\"  The .xt macro is for ``Xresource true'' -- for enabling a
.\"  meter, like the .pm macro/-+foo.
.de xt
xosview*\\$1: (True or False)
.RS
If True then xosview will display a \\$1 meter.
.RE
..
.\"  The .uf macro is for the ``used format'' stuff.
.de uf
xosview*\\$1UsedFormat: (float, percent or autoscale)
.RS
This resource tells xosview how to display "used" labels.  The formats work
as follows:

\fBfloat\fP:
.RS
Display the value as a floating point number.
.RE
\fBpercent\fP:
.RS
Display the value as a percentage of the total.
.RE
\fBautoscale\fP:
.RS
Display the absolute value and automatically print the units (K, M, or G) as
appropriate.
.RE
.RE
..
.\"  Define a color macro for the various xosview*fooXXXColor: resources.
.\"  Usage:  .cc swap Used used "swap space"  <--  keep 'swap
\"                                                       space' as one arg.
.de cc
xosview*\\$1\\$2Color: \fIcolor\fP
.RS
The \\$1 meter will use this color to display the \\$3 field.
.RE
..
.\"  ================  End of Macros  =============================
.SH DESCRIPTION
\fIxosview\fP is a monitor which displays the status of several system
parameters.  These include CPU usage, load average, memory, swap
space, network usage and more.  Each resource is displayed as a
horizontal bar which is separated into color coded regions showing how
much of the resource is being put to a particular use.

\fIxosview\fP runs on several platforms (Linux, NetBSD, FreeBSD,
OpenBSD, DragonflyBSD, some Solaris systems, IRIX 6.5, HPUX, AIX 4.x
and 5.x, and GNU).
Not all of the meters described below are supported on all platforms.
Some of the meters may appear different depending upon the platform
\fIxosview\fP is running on.  Note that *BSD is used as an
abbreviation for all the supported BSD operating systems (NetBSD,
FreeBSD, OpenBSD, DragonflyBSD).

\fBLoad\fP: Kernel reported load average
   all platforms  :  load
   Linux          :  load | CPU frequency
   *BSD           :  load | CPU frequency
   SunOS5         :  load | CPU frequency

\fBCPU Usage\fP
   Linux  : user | nice | system | soft-interrupt | interrupt | io-wait |
            guest | niced guest | stolen | idle
   *BSD   : user | nice | system | interrupt | idle
   SunOS5 : user | system | wait | idle
   IRIX   : user | system | interrupt | wait | idle
   HPUX   : user | nice | system | interrupt | idle
   AIX    : user | system | io-wait | idle

\fBMemory Usage\fP
   Linux  : used | buffers | slab | map | cache | free
   FreeBSD: active | inactive | wired | cache | free
   DFBSD  : active | inactive | wired | cache | free
   OpenBSD: active | inactive | wired | free
   NetBSD : active | inactive | wired | free
   SunOS5 : kernel | zfs | other | free
   IRIX   : kernel | fs | user | free
   HPUX   : text | used | other | free
   AIX    : used | cache | free
   GNU    : active | inactive | wired | free

\fBSwap Usage\fP
   Linux  : used | free
   *BSD   : used | free
   SunOS5 : used | free
   HPUX   : used | free
   AIX    : used | free
   GNU    : used | free

\fBDisk Usage\fP: Throughput to/from local disk, per second
   Linux  : in | out | idle
   *BSD   : in | out | idle
   SunOS5 : in | out | idle
   AIX    : in | out | idle

\fBPage Swapping\fP: Pages to/from swap, per second
   Linux  : in | out | idle
   *BSD   : in | out | idle
   SunOS5 : in | out | idle
   HPUX   : in | out | idle
   AIX    : in | out | idle
   GNU    : in | out | idle

\fBNetwork Usage\fP
   Linux  : in | out | idle
   *BSD   : in | out | idle
   SunOS5 : in | out | idle
   AIX    : in | out | idle

\fBGFX Usage\fP: Framebuffer exchanges, per second
   IRIX   : swapbuffers

\fBInterrupts\fP: "leds" which blink when an interrupt occurs
   Linux  : IRQs
   *BSD   : IRQs

\fBInterrupt Rate\fP: Per second
   Linux  : interrupts | free
   *BSD   : interrupts | free
   SunOS5 : interrupts | free

\fBSerial Port Status\fP: "leds" which show the serial port parameters
   Linux  : LSR and MSR

\fBBattery Level\fP: charge and status of batteries
   Linux  : available | used
   *BSD   : available | used

\fBRAID\fP: State of disks in a software RAID array
   Linux  : disk0 disk1 disk2 ... diskN | rebuild

\fBWireless Link\fP
   Linux  : quality

\fBSensors\fP: Readings from sensors
   Linux  : I2C/hwmon sensors (lmstemp)
            Intel Core/AMD K8+/VIA C7 temperature (coretemp)
            ACPI thermal zones (acpitemp)
   *BSD   : I2C sensors (bsdsensor)
            Intel Core/AMD K8+ temperature (coretemp)

Typing a 'q' in the window will terminate xosview.

.SH OPTIONS

Most of these command line options are just a convenient way to set one or
more of \fIxosview\fP's resources.  Please see the \fBRESOURCES\fP section for
more details on what the resource is for.

\-v
.RS
Displays the version number.
.RE

\-name \fIname\fP
.RS
Sets the Resource name xosview will use (same as the \-name option
for typical X applications like xterm).  When performing resource
lookups, xosview will look under \fIname\fP, and then under
``xosview''.  For an example, perform the following (as
documented in README.netbsd):
.RS
  xrdb \-merge Xdefaults.stipple
  xosview \-name xosvstipple &
  xosview \-name xosvstipplebw &
.RE
.RE

\-display \fIdisplay\fP
.RS
Sets the X display to \fIdisplay\fP.  This option overrides the
xosview*display resource.
.RE

\-font \fIfont\fP
.RS
Specifies the font xosview will use for labels.  This option overrides the
xosview*font resource.
.RE

\-title \fItitle\fP
.RS
This option sets the name xosview will tell the window manager to use for
the X window and icon.  This option overrides the xosview*title resource.
.RE

\-geometry \fIgeometry_string\fP
.RS
Sets the X geometry to \fIgeometry_string\fP.  This option overrides the
xosview*geometry resource.
.RE

.pm captions
.pm labels
.pm usedlabels
.pm cpu

\-cpus
.RS
Force the display of a single meter for all CPUs in the system. This
option is equivalent to setting xosview*cpuFormat to "single".
.RE

\+cpus
.RS
Force the display of all CPUs in the system. This option is equivalent
to setting xosview*cpuFormat to "all".
.RE

.pm load
.pm mem
.pm swap
.pm battery
.pm gfx
.pm wireless
.pm net

\-network \fImaxbandwidth\fP
.RE
\-networkBW \fImaxbandwidth\fP
.RE
\-networkBandwidth \fImaxbandwidth\fP
.RS
These options override the xosview*netBandwidth resource.  They cause
xosview to display a meter that will shows network usage, with a maximum
bandwidth of \fBmaxbandwidth\fP.  Notice that setting the bandwidth to
0 no longer disables the meter \-\- use the ``\-net'' option instead.
.RE

.pm page

\-pagespeed \fIval\fP
.RS
This option overrides the xosview*pageBandWidth resource.  The resource
xosview*pageBandWidth will be set to \fIval\fP.
.RE

.pm disk
.pm int

\-ints +ints
.RE
\-interrupts +interrupts
.RS
Equivalent to \-int and +int.
.RE

.pm irqrate

\-intrate +intrate
.RS
Equivalent to \-irqrate and +irqrate.
.RE

.pm lmstemp
.pm coretemp
.pm acpitemp
.pm bsdsensor
.RE

\-xrm \fIresource_string\fP
.RS
This switch allows any of xosview's resources to be set on the command line.
An example of how the xosview*memFreeColor could be set using this option is
shown below. Note the use of quoting to prevent the shell from expanding '*'
or from creating two separate arguments, 'xosview*memfreeColor:'
and 'purple'):
.RS
\-xrm "xosview*memFreeColor: purple"
.RE
.RE

.SH X RESOURCES

The following is a list of X resources supported by \fIxosview\fP.  Each has
a default value assigned to it.  These values can be found in the file
Xdefaults which can be obtained in the source distribution of xosview.
They can be overridden in the usual places (/usr/lib/X11/app-defaults/XOsview,
$HOME/.Xdefaults, etc.).

It should be noted that it is OK to have a resource defined for a port of
xosview that does not support the feature the resource configures.  Xosview
will simply ignore the resources that are set for it but not supported on
a given platform.


\fBGeneral Resources\fP

xosview*title: \fIname\fP
.RS
The string that xosview will use for the X window title.  Normally xosview
will use 'xosview@machine_name' for a title.  This resource overrides the
default behavior.
.RE

xosview*geometry: \fIgeometry_string\fP
.RS
This is a standard X geometry string that defines the size and location of
the X window used by xosview.
.RE

xosview*display: \fIname\fP
.RS
The name of the display where xosview will contact the X server for drawing
its window.
.RE

xosview*pixmapName: \fIname\fP
.RS
The filename of an X pixmap (xpm) file for use as a background
image.
.RE

xosview*captions: (True or False)
.RS
If True then xosview will display meter captions.
.RE

xosview*labels: (True or False)
.RS
If True then xosview will display meter labels.
.RE

xosview*meterLabelColor: \fIcolor\fP
.RS
The color to use for the meter labels.
.RE

xosview*usedlabels: (True or False)
.RS
If True then xosview will display labels that show the percentage of the
resource (or absolute amount, depending on the meter) being used.  This
option requires that the labels option also be set to True.
.RE

xosview*usedLabelColor: \fIcolor\fP
.RS
The color to use for "used" labels.
.RE

xosview*borderwidth: \fIwidth\fP
.RS
The width of the border for the xosview window.
.RE

xosview*font: \fIfont\fP
.RS
This is the font that xosview will use. This is a bitmap font
rendered at the X server. To list the available fonts, use
.BR xlsfonts (1),
and rehash the list with
.BR xset (1).
.RE

xosview*background: \fIcolor\fP
.RS
This is the color that will be used for the background.
.RE

xosview*foreground: \fIcolor\fP
.RS
This is the color that will be used for the foreground.
.RE

xosview*enableStipple: (True or False)
.RS
Change to true to try stipple support.  This is primarily for users
stuck with 1-bit monitors/display cards.  Try setting enableStipple
true.  Please give us feedback on this, if you use it.  It needs
some more work, but no one has given us any feedback so far.
.RE

xosview*graphNumCols: \fInumber\fP
.RS
This defines the number of sample bars drawn when a meter is in scrolling
graph mode. This also has the side-effect of defining the width of the
graph columns. This is only used by meters which have graph mode enabled.
.RE


\fBLoad Meter Resources\fP

.\"  Do the load: True resource.
.xt load

xosview*loadProcColor: \fIcolor\fP
.RS
This is the color that the load meter will use to display the load average
when it is below the warning threshold.
.RE

xosview*loadWarnColor: \fIcolor\fP
.RS
This is the color that the load meter will use once the load average is
above the warning but below the critical load threshold.
.RE

xosview*loadCritColor: \fIcolor\fP
.RS
This is the color that the load meter will use once the load average is
above critical load threshold.
.RE

.\"  loadIdleColor
.cc load Idle idle

.\"  Do the priority resource
.pp load

xosview*loadWarnThreshold: \fIint\fP
.RS
This number (which must be an integer >= 1) sets the value at which
the loadmeter changes its status and color from "normal" to "warning".
The default value is the number of processors.
.RE

xosview*loadCritThreshold: \fIint\fP
.RS
This number (which must be an integer >= 1) sets the value at which
the loadmeter changes its status and color from "warning" to "critical".
The default value is four times the warning threshold.
.RE

xosview*loadDecay: (True or False)
.RS
You should probably leave this at the default value (False).  The load
is already a time-averaged value!
.RE

.dg load

.\"  loadUsedFormat resource
.uf load

xosview*loadCpuSpeed: (True or False)
.RS
Display the current CPU speed in the load meter.
.RE


\fBCPU Meter Resources\fP

xosview*cpu: (True or False)
.RS
If True then xosview will display a cpu meter. On Linux, *BSD, Solaris and
IRIX SMP machines, the resource cpuFormat defines how meters are created for
multiple CPUs.
.RE

.\"  cpuUserColor, cpuNiceColor, cpuSystemColor, cpuInterruptColor, cpuFreeColor
.cc cpu User "cpu user time"

.cc cpu Nice "cpu nice time"

.cc cpu System "cpu system time"

.cc cpu Interrupt "cpu hard interrupt time"

.cc cpu SInterrupt "cpu soft interrupt time"

.cc cpu Wait "cpu IO waiting time"

.cc cpu Guest "cpu virtualization guest time"

.cc cpu NiceGuest "cpu niced virtualization guest time"

.cc cpu Stolen "cpu involuntary wait time"

.cc cpu Free "cpu idle time"

.\"  Priority, decay, usedFormat resources:
.pp cpu

.dc cpu

.dg cpu

.uf cpu

xosview*cpuFormat: (single, all, both or auto)
.RS
If `single', only a cumulative meter for all CPU usage is created.
`all' creates a meter for each CPU, but no cumulative meter. `both'
creates one cumulative meter and one for each CPU. `auto' makes a
choice based on the number of CPUs found.
.RE

xosview*cpuFields: USED/USR/NIC/SYS/INT/SI/HI/WIO/GST/NGS/STL/IDLE
.RS
The set of fields to show in Linux CPU meter instead of the
default. Possible fields are:

\fBUSED\fP:
.RS
Combine all used CPU time into one field. This is the sum of user, nice,
system, soft and hard interrupts, guest, niced guest and stolen times.
None of these, except stolen, may be defined together with `USED'.
.RE
\fBIDLE\fP:
.RS

CHANGES
For changes since xosview-1.9.1, see the source code history in Git.

Changes since xosview-1.9.0

  - Fix a bug where battery meter was not redrawn

Changes since xosview-1.8.3

  - Modify linux serialmeter startup to allow for a port number instead of
    True/False.  Patch provided by Horst Wente.

  - netIface support for NetBSD.  Patch from  Bernd Ernesti.

  - Move repository to Git

  - Merged patches from Debian distribution

  - Several bugfixes

  - Improvements to defaults

Changes since xosview-1.8.2
 - NFSDStats ans NFSStats Graph, Decay and UsedMode are now configurable

 - SunOS updates, compiles now cleanly with gcc 3.3.2

 - Fix for linux intmeter for interrupts > 16 thanks to John Adams

 - Small patch from Gunnar Wolf for lmstemp (change int to double)

 - Fix battery meter for case where lack of APM report -1%.  Thanks
   to Gunnar Wolf.

 - Allow netmeter to specify, via the Xresource xosview*netIface,
   which network interface it will monitor.  Patch thanks to Gunnar Wolf.

 - Add patch for additional cpu fields in linux for 2.6 kernel.  Patch thanks
   to Hendrik Fehr.

 - Add various pathes from the NetBSD project.  Pathes contributed by Roland
   Illig.

 - Fix bug 854215 (-v and --version still needed to connect to X server)

Changes since xosview-1.8.1
 - Applied sourceforge patch 927112 from Russell Reed which fixes
   the swapmeter when USESYSCALLS is defined.

 - Applied sourceforge patch 959433 which adds acpi support to
   the linux battery meter.

 - Applied a patch similar to the sourceforge patch 586184 submitted
   by Mark Guertin.

 - Linux diskmeter now uses vmstat and thus works with 2.6.
   But the units problem (block size) probably still remains.

 - Linux pagemeter should now work with 2.6 kernels

 - Merge in the fedora-2 nfs patch.

 - Merge in the fedora-2 strip patch.

 - Merge in the fedora-2 proc patch.

 - Merge in the fedora-2 s390 patch.

 - Merge in the fedora-2 non-i386 patch.

 - Merge in the fedora-2 linux-2.6 kernel (related) patch.

 - Add an autogen.sh script.  This script runs autoconf to create
   the configure script.

Changes since xosview-1.8.0

- Update linux port to build with gcc 3.2 and 3.3 (should still work
  with 2.95.

- Add -Wno-deprecated for gcc builds.  This avoids warnings about
  the new C++ header files.  This should be fixed for the next release.

- Change MAX_PROC_LENGTH in linux/cpumeter to 4096 to avoid
  an infinite loop with redhat 7.3

Changes since xosview-1.7.3

- Add Mike Butler's patch which fixes an infinite loop in
  the linux netmeter when a pcmcia card is ejected.

- Add Leopold Toetsch's linux lmstemp meter.

- Linux port (and maybe others) now works with gcc-3.0.4

- added IRIX 6.5 support

- Updated NetBSD support, based on patches from Bernd Ernesti, to
  use new CPU meter sysctl interface.

Changes since xosview-1.7.2

- IRQ rate meter for BSDs added.

- FreeBSD 4.* fixes by David O'Brien

Changes since xosview-1.7.1

- Linux now has a RAID meter.  See README.linux for how to set it up.

  BSDI support, courtesy of Tomer Klainer.
  Fixed memstat module for linux-2.2

  Changes from Thomas Waldmann (ThomasWaldmann@gmx.de)
- Fixed cosmetical bug in xosview:
  if you tried to make the xosview window height too small, it looked quite
  strange (because the height value for the meters got negative)

- Fixed cosmetical bug in disk meter:
  at first call, it displayed negative values, overwriting the "DISK" label

- Cosmetical fix:
  The titles were at x=0 - directly at the window border.
  Fixed it to show 1 pixel background between window border and title texts.
  Also moved "used" displays 2 pixels to the right, so that there is also 1
  pixel between the titles and the used display texts.

- Cosmetical fix: Xdefaults and Xdefaults.stipple file now looks nicer.

- Cosmetical fix: load meter sometimes went beyond the right border (I think
  when showing values >>20). Changed load meter scaling to be really dynamic
  (no upper bound for total, but total doesn't become less than 1.0).
  Changed "used" display to show load relative to 1.0 (so it now just shows
  the load value ;-).

- New feature: load meter now has not only a warning color (and threshold),
  but also a (higher) critical color (and threshold).

- Added info to disk meter problem documentation (see source):
  on my system, it showed only the half of the correct values, factor
  for SW-RAID5 md devices seems to be 2048 (md0), while standard ext2fs
  factor is 1024 on my system, too (hda).

  So there is a problem. Solution ???

- Fixed bitmeter to calculate graphics coordinates like fieldmeter

- MemMeter showed strange values when starting vmware (and having memstat
  module loaded): "used" memory got negative !! Looking at the values, I had
  the impression that "shared" memory is also counted as cache memory.
  I hopefully fixed that by subtracting shared from cache - looks better now.

Changes since xosview-1.7.0
  Optimised scrolling graph to use XCopyArea.
  Fixes to linux diskmeter.
  Fixed all (I hope) linux code which parsed /proc/stat with too
    small a buffer.

Changes since xosview-1.6.2
  Added a linux diskmeter which *should* be similar to the BSD version
    but which also differentiates between reading and writing. Linux
    looks like it actually has more stats (per disk input/output) so
    this could be expanded upon.
  Better *BSD support:  OpenBSD intmeter, FreeBSD diskmeter,
    NetBSD battery meter.
  Some Solaris support.

Changes since xosview-1.6.1
  Check for and set -D__SMP__ if the linux memstat module is built on a
    SMP system.
  New option to omit the meter captions (+-captions) added by
    Christian 'Dr. Disk' Hechelmann <drdisk@tc-gruppe.de>
  New graphing meter display mode added.
  Fixes to work with linux/alpha.
  BSD directories merged -- now a single bsd subdir shared by
    NetBSD, FreeBSD, and OpenBSD.

Changes since xosview-1.6.0:

xosview.cc
//
//  Copyright (c) 1994, 1995, 2002, 2006 by Mike Romberg ( mike.romberg@noaa.gov )
//
//  This file may be distributed under terms of the GPL
//

#include "xosview.h"
#include "meter.h"
#include "MeterMaker.h"
#if ( defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || \
      defined(XOSVIEW_OPENBSD) || defined(XOSVIEW_DFBSD) )
# include "kernel.h"
#endif
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <iostream>

static const char * const versionString = "xosview version: Git";

static const char NAME[] = "xosview@";

#if !defined(__GNUC__)

#define MIN(x,y)		\
(				\
    x < y ? x : y		\
)

#define MAX(x,y)		\
(				\
    x > y ? x : y		\
)

#else

#define MIN(x,y)		\
({				\
    const typeof(x) _x = x;	\
    const typeof(y) _y = y;	\
				\
    (void) (&_x == &_y); 	\
				\
    _x < _y ? _x : _y;		\
})

#define MAX(x,y)		\
({				\
    const typeof(x) _x = x;	\
    const typeof(y) _y = y;	\
				\
    (void) (&_x == &_y); 	\
				\
    _x > _y ? _x : _y;		\
})

#endif // sgi

double MAX_SAMPLES_PER_SECOND = 10;

XOSView::XOSView( const char * instName, int argc, char *argv[] ) : XWin(),
						xrm(Xrm("xosview", instName)){
  // Check for version arguments first.  This allows
  // them to work without the need for a connection
  // to the X server
  checkVersion(argc, argv);

  setDisplayName (xrm.getDisplayName( argc, argv));
  openDisplay();  //  So that the Xrm class can contact the display for its
		  //  default values.
  //  The resources need to be initialized before calling XWinInit, because
  //  XWinInit looks at the geometry resource for its geometry.  BCG
  xrm.loadAndMergeResources (argc, argv, display_);
  XWinInit (argc, argv, NULL, &xrm);
#if 1	//  Don't enable this yet.
  MAX_SAMPLES_PER_SECOND = atof(getResource("samplesPerSec"));
  if (!MAX_SAMPLES_PER_SECOND)
    MAX_SAMPLES_PER_SECOND = 10;
#endif
  usleeptime_ = (unsigned long) (1000000/MAX_SAMPLES_PER_SECOND);
  if (usleeptime_ >= 1000000) {
    /*  The syscall usleep() only takes times less than 1 sec, so
     *  split into a sleep time and a usleep time if needed.  */
    sleeptime_ = usleeptime_ / 1000000;
    usleeptime_ = usleeptime_ % 1000000;
  } else { sleeptime_ = 0; }
#if ( defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || \
      defined(XOSVIEW_OPENBSD) || defined(XOSVIEW_DFBSD) )
  BSDInit();	/*  Needs to be done before processing of -N option.  */
#endif

  hmargin_  = atoi(getResource("horizontalMargin"));
  vmargin_  = atoi(getResource("verticalMargin"));
  vspacing_ = atoi(getResource("verticalSpacing"));
  hmargin_  = MAX(0, hmargin_);
  vmargin_  = MAX(0, vmargin_);
  vspacing_ = MAX(0, vspacing_);

  checkArgs (argc, argv);  //  Check for any other unhandled args.
  xoff_ = hmargin_;
  yoff_ = 0;
  nummeters_ = 0;
  meters_ = NULL;
  name_ = const_cast<char *>("xosview");
  _deferred_resize = true;
  _deferred_redraw = true;
  windowVisibility = OBSCURED;

  //  set up the X events
  addEvent( new Event( this, ConfigureNotify,
		      (EventCallBack)&XOSView::resizeEvent ) );
  addEvent( new Event( this, Expose,
		      (EventCallBack)&XOSView::exposeEvent ) );
  addEvent( new Event( this, KeyPress,
		      (EventCallBack)&XOSView::keyPressEvent ) );
  addEvent( new Event( this, VisibilityNotify,
                      (EventCallBack)&XOSView::visibilityEvent ) );
  addEvent( new Event( this, UnmapNotify,
                      (EventCallBack)&XOSView::unmapEvent ) );

  // add or change the Resources
  MeterMaker mm(this);

  // see if legends are to be used
  checkOverallResources ();

  // add in the meters
  mm.makeMeters();
  for (int i = 1 ; i <= mm.n() ; i++)
    addmeter(mm[i]);

  if (nummeters_ == 0) {
    std::cerr << "No meters were enabled!  Exiting..." << std::endl;
    exit (0);
  }

  //  Have the meters re-check the resources.
  checkMeterResources();

  // determine the width and height of the window then create it
  figureSize();
  init( argc, argv );
  title( winname() );
  iconname( winname() );
  dolegends();
}

void XOSView::checkVersion(int argc, char *argv[]) const
    {
    for (int i = 0 ; i < argc ; i++)
        if (!strncasecmp(argv[i], "-v", 2)
          || !strncasecmp(argv[i], "--version", 10))
            {
            std::cerr << versionString << std::endl;
            exit(0);
            }
    }

void XOSView::figureSize ( void ) {
  if ( legend_ ){
    if ( !usedlabels_ )
      xoff_ = textWidth( "XXXXXX" );
    else
      xoff_ = textWidth( "XXXXXXXXXX" );

    yoff_ = caption_ ? textHeight() + textHeight() / 4 : 0;
  }
  static int firsttime = 1;
  if (firsttime) {
    firsttime = 0;
    width_ = findx();
    height_ = findy();
  }
  else
  {
  }
}

void XOSView::checkMeterResources( void ){
  MeterNode *tmp = meters_;

  while ( tmp != NULL ){
    tmp->meter_->checkResources();
    tmp = tmp->next_;
  }
}

int XOSView::newypos( void ){
  return 15 + 25 * nummeters_;
}

void XOSView::dolegends( void ){
  MeterNode *tmp = meters_;
  while ( tmp != NULL ){
    tmp->meter_->docaptions( caption_ );
    tmp->meter_->dolegends( legend_ );
    tmp->meter_->dousedlegends( usedlabels_ );
    tmp = tmp->next_;
  }
}

void XOSView::addmeter( Meter *fm ){
  MeterNode *tmp = meters_;

  if ( meters_ == NULL )
    meters_ = new MeterNode( fm );
  else {
    while ( tmp->next_ != NULL )
      tmp = tmp->next_;
    tmp->next_ = new MeterNode( fm );
  }
  nummeters_++;
}

int XOSView::findx( void ){
  if ( legend_ ){
    if ( !usedlabels_ )
      return textWidth( "XXXXXXXXXXXXXXXXXXXXXXXX" );
    else
      return textWidth( "XXXXXXXXXXXXXXXXXXXXXXXXXXXXX" );
  }
  return 80;
}

int XOSView::findy( void ){
  if ( legend_ )
    return 10 + textHeight() * nummeters_ * ( caption_ ? 2 : 1 );

  return 15 * nummeters_;
}

void XOSView::checkOverallResources() {
  //  Check various resource values.

  //  Set 'off' value.  This is not necessarily a default value --
  //    the value in the defaultXResourceString is the default value.
  usedlabels_ = legend_ = caption_ = 0;

  setFont();

   // use captions
  if ( isResourceTrue("captions") )
      caption_ = 1;

  // use labels
  if ( isResourceTrue("labels") )
      legend_ = 1;

  // use "free" labels
  if ( isResourceTrue("usedlabels") )
      usedlabels_ = 1;
}

const char *XOSView::winname( void ){
  char host[100];
  gethostname( host, 99 );
  host[99] = 0;
  static char name[120];    /*  We return a pointer to this,
                                so it can't be local.  */
  snprintf( name, 120, "%s%s", NAME, host);
  //  Allow overriding of this name through the -title option.
  return getResourceOrUseDefault ("title", name);
}


void  XOSView::resize( void ){
  int spacing = vspacing_+1;
  int topmargin = vmargin_;
  int rightmargin = hmargin_;
  int newwidth = width_ - xoff_ - rightmargin;

  int newheight =
        (height_ -
         (topmargin + topmargin + (nummeters_-1)*spacing + nummeters_*yoff_)
        ) / nummeters_;
  newheight = (newheight >= 2) ? newheight : 2;


  int counter = 1;
  MeterNode *tmp = meters_;
  while ( tmp != NULL ) {
    tmp->meter_->resize( xoff_,
                         topmargin + counter*yoff_ + (counter-1)*(newheight+spacing),
                        newwidth, newheight );
    tmp = tmp->next_;

    counter++;
  }
}


XOSView::~XOSView( void ){
  MeterNode *tmp = meters_;
  while ( tmp != NULL ){
    MeterNode *save = tmp->next_;
    delete tmp->meter_;
    delete tmp;
    tmp = save;
  }
}

void XOSView::draw(void) {
  if (windowVisibility != OBSCURED) {
    MeterNode *tmp = meters_;

    XOSDEBUG("Doing draw.\n");
    clear();

    while (tmp != NULL) {
      tmp->meter_->draw();
      tmp = tmp->next_;
    }
  }
  else {
    XOSDEBUG("Skipping draw:  not visible.\n");
  }
}

void XOSView::run( void ){
  while(!done_) {
    // Check for X11 events
    checkevent();

    // Check if the window has been resized (at least once)
    if (_deferred_resize) {
      resize();
      _deferred_resize = false;
      _deferred_redraw = true;
    }

    // redraw everything if needed
    if (_deferred_redraw) {
      draw();
      _deferred_redraw = false;
    }

    // Update the metrics & meters
    MeterNode *tmp = meters_;
    while ( tmp != NULL ){
      if ( tmp->meter_->requestevent() )
        tmp->meter_->checkevent();
      tmp = tmp->next_;
    }

    flush();

    /*  First, sleep for the proper integral number of seconds --
     *  usleep only deals with times less than 1 sec.  */
    if (sleeptime_) sleep((unsigned int)sleeptime_);
    if (usleeptime_) usleep( (unsigned int)usleeptime_);
  }
}

void XOSView::keyPressEvent( XKeyEvent &event ){
  char c = 0;
  KeySym key;

  XLookupString( &event, &c, 1, &key, NULL );

  if ( (c == 'q') || (c == 'Q') )
    done_ = 1;
}

void XOSView::checkArgs (int argc, char** argv) const
{
  //  The XWin constructor call in the XOSView constructor above
  //  modifies argc and argv, so by this
  //  point, all XResource arguments should be removed.  Since we currently
  //  don't have any other command-line arguments, perform a check here
  //  to make sure we don't get any more.
  if (argc == 1) return;  //  No arguments besides X resources.

  //  Skip to the first real argument.
  argc--;
  argv++;
  while (argc > 0 && argv && *argv)
  {
    switch (argv[0][1]) {
      case 'n': //  Check for -name option that was already parsed
		//  and acted upon by main().
		if (!strncasecmp(*argv, "-name", 6))
		{
		  argv++;	//  Skip arg to -name.
		  argc--;
		}
		break;
#if ( defined(XOSVIEW_NETBSD) || defined(XOSVIEW_FREEBSD) || \
      defined(XOSVIEW_OPENBSD) || defined(XOSVIEW_DFBSD) )
      case 'N': if (strlen(argv[0]) > 2)
      		  SetKernelName(argv[0]+2);
		else
		{
		  SetKernelName(argv[1]);
		  argc--;
		  argv++;
		}
		break;
#endif
	      /*  Fall through to default/error case.  */
      default:
      		std::cerr << "Ignoring unknown option '" << argv[0] << "'.\n";
	  	break;
    }
    argc--;
    argv++;
  }
}

void XOSView::exposeEvent(XExposeEvent &event) {
  _deferred_redraw = true;
  XOSDEBUG("Got expose event.\n");
}

/*
 * All window changes come in via XConfigureEvent (not
 * XResizeRequestEvent)
 */

void XOSView::resizeEvent( XConfigureEvent &event ) {
  XOSDEBUG("Got configure event.\n");

  if (event.width == width_ && event.height == height_)
    return;

  XOSDEBUG("Window has resized\n");

  width(event.width);
  height(event.height);

  _deferred_resize = true;
}

void XOSView::visibilityEvent(XVisibilityEvent &event) {
  if (event.state == VisibilityPartiallyObscured) {
    if (windowVisibility != FULLY_VISIBLE)
      _deferred_redraw = true;
    windowVisibility = PARTIALLY_VISIBILE;
  }
  else if (event.state == VisibilityFullyObscured) {
    windowVisibility = OBSCURED;
    _deferred_redraw = false;
  }
  else {
    if (windowVisibility != FULLY_VISIBLE)
      _deferred_redraw = true;
    windowVisibility = FULLY_VISIBLE;
  }

  XOSDEBUG("Got visibility event: %s\n",
    (windowVisibility == FULLY_VISIBLE)
        ? "Full"
        : (windowVisibility == PARTIALLY_VISIBILE)
            ? "Partial"
            : "Obscured"
    );
}

void XOSView::unmapEvent( XUnmapEvent & ev ){
  /* unclutter creates a subwindow of our window if it hides the cursor,
     we get the unmap event if the cursor is moved again. Don't treat it
     as main window unmap */
  if(ev.window == window_)
    windowVisibility = OBSCURED;
}
